pkg/storage/ingest/reader_test.go

```diff
@@ -87,6 +89,155 @@ func TestReader_ConsumerError(t *testing.T) {
+	})
+
+	t.Run("should block until the context deadline exceed if produced records are not consumed", func(t *testing.T) {
+		t.Parallel()
+
+		consumer, reader, writeClient, reg := setup(t)
+
+		// Produce some records.
+		produceRecord(ctx, t, writeClient, topicName, partitionID, []byte("record-1"))
+		t.Log("produced 1 record")
+
+		err := reader.WaitReadConsistency(createTestContextWithTimeout(t, time.Second))
+		require.ErrorIs(t, err, context.DeadlineExceeded)
+
+		// Consume the records.
+		records, err := consumer.waitRecords(1, time.Second, 0)
+		assert.NoError(t, err)
+		assert.Equal(t, [][]byte{[]byte("record-1")}, records)
+
+		// Now the WaitReadConsistency() should return soon.
+		err = reader.WaitReadConsistency(createTestContextWithTimeout(t, time.Second))
+		require.NoError(t, err)
+
+		assert.NoError(t, promtest.GatherAndCompare(reg, strings.NewReader(`
+			# HELP cortex_ingest_storage_strong_consistency_requests_total Total number of requests for which strong consistency has been requested.
+			# TYPE cortex_ingest_storage_strong_consistency_requests_total counter
+			cortex_ingest_storage_strong_consistency_requests_total 2
+
+			# HELP cortex_ingest_storage_strong_consistency_failures_total Total number of failures while waiting for strong consistency to be enforced.
+			# TYPE cortex_ingest_storage_strong_consistency_failures_total counter
+			cortex_ingest_storage_strong_consistency_failures_total 1
+		`), "cortex_ingest_storage_strong_consistency_requests_total", "cortex_ingest_storage_strong_consistency_failures_total"))
+	})
+
+	t.Run("should return if no records have been produced yet", func(t *testing.T) {
+		t.Parallel()
+
+		_, reader, _, reg := setup(t)
+
+		err := reader.WaitReadConsistency(createTestContextWithTimeout(t, time.Second))
+		require.NoError(t, err)
+
+		assert.NoError(t, promtest.GatherAndCompare(reg, strings.NewReader(`
+			# HELP cortex_ingest_storage_strong_consistency_requests_total Total number of requests for which strong consistency has been requested.
+			# TYPE cortex_ingest_storage_strong_consistency_requests_total counter
+			cortex_ingest_storage_strong_consistency_requests_total 1
+
+			# HELP cortex_ingest_storage_strong_consistency_failures_total Total number of failures while waiting for strong consistency to be enforced.
+			# TYPE cortex_ingest_storage_strong_consistency_failures_total counter
+			cortex_ingest_storage_strong_consistency_failures_total 0
+		`), "cortex_ingest_storage_strong_consistency_requests_total", "cortex_ingest_storage_strong_consistency_failures_total"))
+	})
+
+	t.Run("should return an error if the PartitionReader is not running", func(t *testing.T) {
+		t.Parallel()
+
+		_, reader, _, reg := setup(t)
+
+		require.NoError(t, services.StopAndAwaitTerminated(ctx, reader))
+
+		err := reader.WaitReadConsistency(createTestContextWithTimeout(t, time.Second))
+		require.ErrorContains(t, err, "partition reader service is not running")
```

Comment:
should we use the global error value here?

Reply:
We don't have a specific one for the `partitionReader`. Reason why I haven't introduced it was cause the current error is dynamic, because it also reports the current service state.